overlord/snapshotstate/backend/backend.go

```diff
@@ -445,6 +446,129 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 	return nil
 }
 
+// Import a snapshot from the export file format
+func Import(ctx context.Context, id uint64, r io.Reader) (size int64, snapNames []string, err error) {
+	comment := fmt.Sprintf("snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	p := path.Join(dirs.SnapCacheDir, "snapshots", fmt.Sprintf("import-%d", id))
+	if err := os.MkdirAll(path.Join(dirs.SnapCacheDir, "snapshots"), 0755); err != nil {
+		return 0, nil, fmt.Errorf("cannot create snapshots directory")
+	}
+	if err := os.Mkdir(p, 0755); err != nil {
+		if os.IsExist(err) {
+			return 0, nil, fmt.Errorf("snapshot import already in progress for ID `%d`", id)
+		}
+		return 0, nil, fmt.Errorf("failed creating import cache: %v", err)
+	}
+	defer os.RemoveAll(p)
+
+	// unpack the tar file to a temporary location (so the contents can be validated)
+	exportFound, size, err := unpackSnapshotImport(r, p)
+	if err != nil {
+		return 0, nil, err
+	}
+	if !exportFound {
+		return 0, nil, fmt.Errorf("snapshot import file incomplete: no export.json file")
+	}
+
+	// walk the cache directory to store the files
+	dir, err := osOpen(p)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed opening import cache for %s: %v", comment, err)
+	}
+	defer dir.Close()
+
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if len(names) > 0 {
+			// move the files into place with the new local set ID
+			names, err := moveCachedSnapshots(names, id, p)
```

Comment:
This needs to undo the partial imported files on failure. i.e. if moveCachedSnapshot files fails after the 2nd imported snapshot file right now we don't cleanup the first imported file AFAICT